feat: add frequency selector for ZapPlanner payments (#1312)

NodeDiver opened the pull request:
## What’s new
- Added frequency selector to the ZapPlanner recurring payment form
- Users can now choose the number and unit (`days`, `weeks`, `months`) for scheduling payments

## Why
- Improves parity with [zapplanner.albylabs.com](https://zapplanner.albylabs.com)
- Solves feature request in #1312

### 🔍 How to test
- Go to Internal Apps → ZapPlanner
- Click "New Recurring Payment"
- Submit to create a new scheduled app + subscription with custom frequency

---

Fixes #1312

reneaaron commented on frontend/src/screens/internal-apps/ZapPlanner.tsx:
```diff
@@ -87,6 +87,8 @@ export function ZapPlanner() {
   const [amount, setAmount] = React.useState("");
   const [comment, setComment] = React.useState("");
   const [senderName, setSenderName] = React.useState("");
+  const [frequencyValue, setFrequencyValue] = React.useState("1");
```
This means by default the payment is done every day. Is that a good default you think? I'd argue monthly payments are probably more common, also that was the default up until now.

reneaaron reviewed (changes requested):
![image](https://github.com/user-attachments/assets/6c64ee48-72ce-44e3-86f0-5deab7908641)

The UI of this select looks wrong: 
- Try to move the "Repeat payment every" below the inputs and make it secondary text color (text-muted-foreground) 
- Use the shadcn select instead of the HTML select to apply proper styles that match other forms
- Make the input wider so more than 1 digit fits (if I want to create a payment every 12 weeks)

NodeDiver commented:
_In short: **Added frequency and USD/BTC selector**_

_List of changes:_
- Added frequency selector: number + unit (days/weeks/months)
- Added USD/BTC toggle with inline conversion preview (using `fiat` tools)
- Updated default values: amount resets to “5” (USD) and frequency to 1 month; BTC option now labeled “BTC (sats)”
- Expanded input boxes for consistency and full visibility of lightning address 
- Calculated dynamic monthly budget cap:  `maxAmount = ceil((sats * periodsPerMonth * 1.01 + 10) * 1.3)`
- Updated `handleSubmit` to always use `satsToSend`and removed legacy `parsedAmount`  
- Converted “months” → days (1 month = 31 days) since month as a unit of time is not compatible (week and days are)
- Fixed dialog‐reopen bug by adding `open` to conversion `useEffect` dependencies

_I'm happy with the results so far 😊, but I know there is room for improvement and more changes. I’m all ears for any suggestions or additional ideas 🙏._

rolznz commented on frontend/src/screens/internal-apps/ZapPlanner.tsx:
```diff
@@ -87,18 +94,68 @@ export function ZapPlanner() {
   const [senderName, setSenderName] = React.useState("");
+  const [frequencyValue, setFrequencyValue] = React.useState("1");
+  const [frequencyUnit, setFrequencyUnit] = React.useState("months");
+  const [amountCurrency, setAmountCurrency] = React.useState<"BTC" | "USD">(
```
Later we could create a `useCurrencies` hook. The code for fetching currencies is currently in `frontend/src/screens/settings/Settings.tsx`. CC @pavanjoshi914

rolznz replied:
This should be a string, not "BTC" | "USD" otherwise we get a type error

NodeDiver replied:
Things done in my latest commit:
- Added support for multicurrency
- 2xbtc entries were shown on the list, I deduplicated BTC entry, only 1 btc on the top.
- Every currency gets converted to sats in the displayed sats preview; 
- Every currency gets converted to sats; in the nwc subscription

I have taken inspiration from the settings and the original zapplanner repo.

rolznz commented on frontend/src/screens/internal-apps/ZapPlanner.tsx:
```diff
@@ -141,14 +224,17 @@ export function ZapPlanner() {
-            amount: parsedAmount,
+            recipientLightningAddress,
+            amount: satsToSend, // ← always sats
+            // TODO: include other currencies
```
currency needs to be provided here right?

rolznz commented on frontend/src/screens/internal-apps/ZapPlanner.tsx:
```diff
@@ -109,13 +166,32 @@ export function ZapPlanner() {
+          throw new Error("Invalid amount");
+        })();
+      // Determine how many payments in one month
+      let periodsPerMonth: number;
```
we should do Math.ceil() on this otherwise if you select every 2 months it will set the value to 0.5 and this will break the budgeting

NodeDiver replied:
Added this line just below the   periodsPerMonth definition,
`    periodsPerMonth = Math.ceil(periodsPerMonth);`

rolznz commented on frontend/src/screens/internal-apps/ZapPlanner.tsx:
```diff
@@ -256,20 +342,85 @@ export function ZapPlanner() {
+                            min="1"
+                            value={frequencyValue}
+                            onChange={(e) => {
+                              const value = Math.max(
```
This creates poor UX e.g. you can't even clear the input. We shouldn't modify the value here at all. Any manipulation of the value can be done in the submit function

NodeDiver replied:
Standardized both inputs to prevent wrong characters/negative numbers:

```
       <Input
                            id="frequency"
                            type="number"
                            min="1"
                            step="1"
                            inputMode="numeric"
                            value={frequencyValue}
                            onChange={(e) => setFrequencyValue(e.target.value)}
                            className="col-span-3 w-70"
                          />
```

rolznz commented on frontend/src/screens/internal-apps/ZapPlanner.tsx:
```diff
@@ -141,14 +224,17 @@ export function ZapPlanner() {
-            recipientLightningAddress: recipientLightningAddress,
-            amount: parsedAmount,
+            recipientLightningAddress,
+            amount: satsToSend, // ← always sats
```
this is not correct, for fiat you provide the fiat value and fiat currency. ZapPlanner will calculate the sat value for every single payment, therefore the payment amount will be "stable" in fiat terms (which is the whole point of adding this feature)

rolznz commented on frontend/src/screens/internal-apps/ZapPlanner.tsx:
```diff
@@ -87,18 +94,68 @@ export function ZapPlanner() {
   const [comment, setComment] = React.useState("");
   const [senderName, setSenderName] = React.useState("");
+  const [frequencyValue, setFrequencyValue] = React.useState("1");
+  const [frequencyUnit, setFrequencyUnit] = React.useState("months");
```
Can you double check the new fields get properly reset when the modal is closed?

NodeDiver replied:
Added this to reset currency selected and frecuency:

```
      setFrequencyValue("1");
      setFrequencyUnit("months");
      setAmountCurrency("USD");
```

rolznz commented on frontend/src/screens/internal-apps/ZapPlanner.tsx:
```diff
@@ -87,35 +94,134 @@ export function ZapPlanner() {
+        const data: Record<string, { name: string }> = await res.json();
+        const fiatCodes = Object.keys(data)
+          .map((c) => c.toUpperCase())
+          // drop "BTC" so it only lives in our prepended slot
```
I don't think our API is supposed to return BTC. So we might be able to remove this line

rolznz commented on frontend/src/screens/internal-apps/ZapPlanner.tsx:
```diff
@@ -87,35 +94,134 @@ export function ZapPlanner() {
+          // drop "BTC" so it only lives in our prepended slot
+          .filter((code) => code !== "BTC")
+          .sort();
+        setCurrencies(["BTC", ...fiatCodes]);
```
It might be weird, but ZapPlanner expects "SATS" as the currency for Bitcoin (see "for developers" on the ZapPlanner page). It seems to return a 400 right now if you set BTC as the currency.

rolznz commented on frontend/src/screens/internal-apps/ZapPlanner.tsx:
```diff
@@ -87,35 +94,134 @@ export function ZapPlanner() {
   const [senderName, setSenderName] = React.useState("");
+  const [frequencyValue, setFrequencyValue] = React.useState("1");
+  const [frequencyUnit, setFrequencyUnit] = React.useState("months");
+  const [amountCurrency, setAmountCurrency] = React.useState<"BTC" | "USD">(
```
could this be renamed to just "currency" instead of "amountCurrency" ?

rolznz commented on frontend/src/screens/internal-apps/ZapPlanner.tsx:
```diff
@@ -131,6 +237,34 @@ export function ZapPlanner() {
+      if (amountCurrency === "BTC") {
+        // user entered sats directly
+        subscriptionBody.amount = satsToSend;
+        subscriptionBody.currency = "BTC";
```
maybe just this line can be removed, so it uses the default currency ("SATS")

rolznz commented on frontend/src/screens/internal-apps/ZapPlanner.tsx:
```diff
@@ -131,6 +237,34 @@ export function ZapPlanner() {
+        subscriptionBody.currency = "BTC";
+      } else {
+        // user entered fiat → ZapPlanner will convert to sats later
+        subscriptionBody.amount = parseFloat(amount);
```
I think unfortunately only whole numbers are supported right now. I tried 0.5 USD and it returned a 400 error.

rolznz reviewed (approved):
tACK
